### DISCUSSION TITLE
with vcpkg install opentelemetry-cpp[otlp_http], no  libopentelemetry_exporter_otlp_http.a or exporters/otlp/*.h

### DISCUSSION THREAD

**Discussion by jodavaho** [open-telemetry/opentelemetry-cpp#1979] [Q&A]
**Describe your environment** 

Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.5 LTS
Release:	20.04
Codename:	focal

vcpkg package management program version 2023-01-24-8a88d63f241d391772fbde69af9cab96c3c64c75

I am compiling *just* the http_main.cc file from `examples/otlp/http_main.cc` in a *new project* that uses vcpkg to satisfy dependencies, to test the vcpkg route to see if that works better than building/installing from source. 

vcpkg.json:

```
{
  "name": "test-otel",
    "version-string": "0.1.0",
    "dependencies": [
    {
      "name": "opentelemetry-cpp",
      "default-features": true,
      "features":["otlp-http"]
    },
    "abseil",
    "protobuf"
    ]
}

```

**Steps to reproduce**
```bash
mkdir build && cd build
cp ~/<path to opentelemetry-cpp>/examples/otlp/http_main.cc ../src/main.cpp
#... make any old ../CMakeLists.txt that has
# find_package(opentelemetry-cpp CONFIG REQUIRED)
cmake .. -DCMAKE_TOOLCHAIN_FILE=~/<path to vcpkg>/scripts/buildsystems/vcpkg.cmake
... vcpkg does its thing ...
$ tree ../vcpkg_installed/x64-linux/include/opentelemetry/exporters/
../vcpkg_installed/x64-linux/include/opentelemetry/exporters/
├── memory
│   ├── in_memory_data.h
│   ├── in_memory_span_data.h
│   ├── in_memory_span_exporter_factory.h
│   └── in_memory_span_exporter.h
└── ostream
    ├── common_utils.h
    ├── log_record_exporter.h
    ├── metric_exporter.h
    ├── span_exporter_factory.h
    └── span_exporter.h
```

**What is the expected behavior?**
```
$ tree ../vcpkg_installed/x64-linux/include/opentelemetry/exporters/
../vcpkg_installed/x64-linux/include/opentelemetry/exporters/
├── otlp
│   └── otlp_http_exporter.h (etc)
├── memory
│   ├── in_memory_data.h
│   ├── in_memory_span_data.h
│   ├── in_memory_span_exporter_factory.h
│   └── in_memory_span_exporter.h
└── ostream
    ├── common_utils.h
    ├── log_record_exporter.h
    ├── metric_exporter.h
    ├── span_exporter_factory.h
    └── span_exporter.h
```

**What is the actual behavior?**

* no otlp_http related headers installed
* no libopentelemetry_exporter_otlp_http.a (or .so)

**Additional context**
Similar to #1975 , I suspect the exporter will not be built unless I also include `opentelemetry-cpp[otlp,otlp_http]`, but that is highly undesirable because it pulls `grpc`, which is currently a showstopper for various reasons. Since `grpc` is not actually required, there should be a mutually exclusive option for http.

**Comment by jodavaho** [comment]
Side question from ignorance: Without exporters and headers, what am I actually pulling when I ask for `otlp_http`?

**Comment by lalitb** [comment]
Moving this to discussion, as the otel-cpp vcpkg is not officially distributed by OpenTelemetry C++.  @ThomsonTan has created the latest vcpkg port v1.8.2, so he may be able to help here.

> Without exporters and headers, what am I actually pulling when I ask for otlp_http?

It should bring exporter and it's headers.

**Reply by jodavaho** [reply]
OK. I suspect the root issue is #1975 and how it requires 3 flags to build `otlp_http`

**Reply by lalitb** [reply]
That shouldn't be the case with v1.8.2, as now it requires only one flag WITH_OTLP_HTTP=ON to build otlp-http. This was fixed for otel-cpp v.1.8.2 in this PR - https://github.com/open-telemetry/opentelemetry-cpp/pull/1916

**Reply by jodavaho** [reply]
Hm. I'm building `main 6d9c4cac Fix #1970 (#1971)`
6d9c4cacc945df970313bf8d48fee183ab2f27b3

**Reply by lalitb** [reply]
Oh my bad. We still need #1926 to get fixed, before this cleanup works properly. If vcpkg is broken for otlp-http, I don't know if Tom will have time to fix this before next release. Please feel free to raise a bug on vcpkg repo, and assign it to Tom.

**Reply by ThomsonTan** [reply]
I just tried the latest vcpkg with opentelemetry 1.8.2 (I did it on Windows, but assume Linux should have the same behavior). And grpc is not installed as dependency, so `opentelemetry-cpp[otlp-http]` only installs HTTP dependency. Is this the ask here?

```cmd
vcpkg install opentelemetry-cpp[otlp-http]:x64-windows-static
dir /s/b otlp_http_exporter*
C:\gh\vcpkg\buildtrees\opentelemetry-cpp\src\v1.8.2-19a2a784eb.clean\exporters\otlp\include\opentelemetry\exporters\otlp\otlp_http_exporter.h
```

**Reply by jodavaho** [reply]
optenelemetry-cpp[otlp-http] does not install headers or libraries for the otlp_http functionality. 

Try:

```
$ tree ../vcpkg_installed/x64-linux/include/opentelemetry/exporters/
../vcpkg_installed/x64-linux/include/opentelemetry/exporters/
├── otlp
│   └── otlp_http_exporter.h (< -- not present!)
├── memory
│   ├── in_memory_data.h
│   ├── in_memory_span_data.h
│   ├── in_memory_span_exporter_factory.h
│   └── in_memory_span_exporter.h
└── ostream
    ├── common_utils.h
    ├── log_record_exporter.h
    ├── metric_exporter.h
    ├── span_exporter_factory.h
    └── span_exporter.h
```

**Reply by jodavaho** [reply]
`opentelemetry-cpp[otlp-http] ` is syntactically _equivalent_ to `WITH_OTLP_HTTP=On`, but that essentially provides _nothing_. The correct flags to get the http exporter without grpc is _for some reason_ `WITH_OTLP=On WITH_OTLP_HTTP=On WITH_OTLP_GRPC=Off` and that won't change, apparently. #1975 #1978

But there is no way to specify those three flags to `vcpkg` just get `otlp-http`, which is the pretty-obvious requirement when specifying `opentelemetry-cpp[otlp-http]`

**Reply by willie** [reply]
Did you ever find a workaround?
